### ISSUE TITLE
Could we also support promises?

### ISSUE THREAD

**Issue by spacesuitdiver** [klis87/redux-requests#244]
I really enjoy using this library and I feel like we could also use it for promises?

(FSA) Payload would look something like:
```
{
  payload: {
    promise: new Promise(...),
  }
}
```

**Comment by klis87** [comment]
Can you show an example and use case?

**Comment by spacesuitdiver** [comment]
The rationale for it is we are looking for a clear way to consistently do error/loading/abort of actions and sometimes it's a request and sometimes it's a promise call to some third party sdk. I'm kind of making this up but something like this:

```
{
  payload: {
    promise: PushWoosh.registerIdentifiedUser(user)
  }
} 
```

**Comment by klis87** [comment]
Interesting use case. If I understand correctly, thats the perfect case for writing a custom driver. I guess below would work:
```js
export const createDriver = () => ({
  requestInstance: null,
  getAbortSource() {
    return null;
  },
  abortRequest() {},
  sendRequest(requestConfig) {
    return requestConfig;
  },
  getSuccessPayload(response) {
    return response;
  },
  getErrorPayload(error) {
    return error;
  },
});
```

Compare it with https://github.com/klis87/redux-saga-requests/blob/master/packages/redux-saga-requests-axios/src/axios-driver.js or https://github.com/klis87/redux-saga-requests/blob/master/packages/redux-saga-requests-fetch/src/fetch-api-driver.js or https://github.com/klis87/redux-saga-requests/blob/master/packages/redux-saga-requests-mock/src/mock-driver.js

Then you could write your request actions like:
```js
{
  payload: {
    request: PushWoosh.registerIdentifiedUser(user)
  }
} 
```

Probably you would have multiple drivers, you can use them like https://github.com/klis87/redux-saga-requests#multiple-drivers-arrow_up

**Comment by spacesuitdiver** [comment]
@klis87 thanks for your response. I actually ended up writing the promise equivalent as mentioned above to this library as an exercise. That said I tend to like your driver approach now that I know it gives an interface to abort requests.

**Comment by klis87** [comment]
Replaced `label` to `enhancement`, as I will add official promise driver at some point. Then it will be possible to use external APIs as well like Firebase SDK.

**Comment by spacesuitdiver** [comment]
@klis87 nice, I'll take a look, I ended up writing my own redux-saga-promise library and it seems to work well though the only iffy part is we pass the promise along with the action payload which just seems to make understanding the action payload when debugging a little tricky is all. It's actually really fun to use because you can pass whatever the heck you want and we use bluebird promise so they are cancellable.

**Comment by klis87** [comment]
@spacesuitdiver only that then action is not serializable, but as long as whole redux state is serializable, I do not see issue with it. Especially now even without promise middleware actions are not serializable anyway if you use for instance `action.meta.getData` with new `networkReducer`

**Comment by strdr4605** [comment]
@spacesuitdiver, can you show redux-saga-promise library you wrote?! As I also tried to write for my case but did not succeed  #270 
I will rethink my implementation when I will have time and update in this issue if found a worthy solution!

**Comment by klis87** [comment]
@strdr4605 i have a suspition, that it could not work due the way i recognize requests actions, I will try to check it today

**Comment by klis87** [comment]
@strdr4605 i am referring to https://github.com/klis87/redux-saga-requests/blob/master/packages/redux-saga-requests/src/actions.js#L65, those checkes need to be quite strict not to conflict other libraries. 

For now I guess u could try this:
```js
{
  type: `FETCH_STH`,
  request: { query: aPromise },
};

export const createDriver = () => ({
  requestInstance: null,
  getAbortSource() {
    return null;
  },
  abortRequest() {},
  sendRequest(requestConfig) {
    return requestConfig.query;
  },
  getSuccessPayload(response) {
    return response;
  },
  getErrorPayload(error) {
    return error;
  },
});
```

**Comment by strdr4605** [comment]
# Posible solution (for "lost souls")

So after trying different approaches this solution works for me:
```ts
export function createRequestsActions(env: EnvInterface): any {
  const placePickerFacade: PlacePickerFacade = new PlacePickerFacade({
    axiosConfig: {
      baseURL: env.PLACE_PICKER_BASE_URL,
    },
  });

// TODO: Track https://github.com/klis87/redux-saga-requests/issues/244 for future changes
  const requestGet: any = (merchantId: string, sessionId: string): object => ({
    type: LocationsActionEnum.NEW_API_LOCATIONS_GET_REQUEST,
    payload: {
      request: {
        query: placePickerFacade.getStoreLocations(merchantId, sessionId),
      },
    },
  });

  return {
    requestGet,
  };
}

// TODO: Track https://github.com/klis87/redux-saga-requests/issues/244 for future changes
export const createDriver: () => Driver = (): Driver => ({
  requestInstance: null,
  getAbortSource(): null {
    return null;
  },
  abortRequest(): void {},
  async sendRequest(requestConfig: { query: Promise<StoreLocationServerResponse> }): Promise<StoreLocationServerResponse> {
    return requestConfig.query;
  },
  getSuccessPayload(response: StoreLocationServerResponse): StoreLocationServerResponse {
    return response;
  },
  getErrorPayload(error: Error): Error {
    console.log({ createDriverError: error });

    return error;
  },
});

export const locationsReducerNew: Reducer = requestsReducer({ actionType: LocationsActionEnum.NEW_API_LOCATIONS_GET_REQUEST });
```
**In my case actions are created dynamically. To have access to them in components I create a context:**
```ts
const theme: DefaultTheme = defaultTheme;
export const { store, requestActions }: any = createAppStore(env, AsyncStorage);
export const RequestActionsContext: React.Context<any> = React.createContext(requestActions);

export const App: React.FC = (): React.ReactElement => (
  <RequestActionsContext.Provider value={requestActions}>
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <Main />
      </Provider>
    </ThemeProvider>
  </RequestActionsContext.Provider>
);
```
and use them in components:
```ts
export const LocationsScreenComponent: React.FC<Props & StateToProps & DispatchToProps> = ({
  dispatch,
  merchantId,
}: Props & StateToProps & DispatchToProps): React.ReactElement => {
  // Get actions from Context
  const requestActions: any = useContext(RequestActionsContext);
  const newGetLocations: any = requestActions.requestGet;

  useEffect(() => {
    if (merchantId) {
      dispatch(newGetLocations(merchantId));
    }
  }, [merchantId]);

  return (
   ...
  );
};

const mapStateToProps: (state: RootState) => StateToProps = ({ merchant }: RootState): StateToProps => ({
  merchantId: merchant.merchantId,
});

export const LocationsScreen: any = connect<StateToProps, {}, DispatchToProps, RootState>(
  mapStateToProps,
)(LocationsScreenComponent);
```
This sample uses [Default dispatch as Prop](https://react-redux.js.org/using-react-redux/connect-mapdispatch#default-dispatch-as-a-prop) but if you have own actions use [bindActionCreators](https://react-redux.js.org/5.x/using-react-redux/connect-mapdispatch#manually-injecting-dispatch)

Looking forward to the implementation of 
```js
{
  type: "items/GET_REQUEST"
  payload: {
    promise: new Promise(...),
  }
}
```
@klis87, let me know if you need some help on this.

**Comment by strdr4605** [comment]
@klis87, as I investigated the codebase I think there is a possibility to add  
```js
{
  type: "items/GET_REQUEST"
  payload: {
    promise: new Promise(...),
  }
}
```
without a need for a separate driver. I think it is possible to add [here](https://github.com/klis87/redux-saga-requests/blob/master/packages/redux-saga-requests/src/sagas/send-request.js#L83)

```js
...
} else if (actionPayload.promise) {
        response = yield call(actionPayload.promise);
    } else if (!Array.isArray(actionPayload.request)) {
...
```
and also check for promise in [isRequestAction](https://github.com/klis87/redux-saga-requests/blob/05ec3552dea3563d5c3e22f822c1cddbcf2659da/packages/redux-saga-requests/src/actions.js#L73).

**Comment by klis87** [comment]
@strdr4605 I want request action to always contain `request` key for consistent object structure. Promise driver is not an issue and I will add it in the future release. I didn't do it yet though as I am working on very big release with simplified API and much more features. Also, drivers will become just functions returning promises thats why I wont work on it in current master branch because I would need to rewrite it anyway soon.

**Comment by klis87** [comment]
Closed in `0.28.0`, thx @strdr4605 ! btw, I added some small option  `processResponse` and only requiring AbortController if someone would like to pass a polyfill
